File: Aegis/Assets/Scripts/MoveBullet.js
#pragma strict
/*
* Script for pushing our linear bullets. 
*/
private var speed : float;
private var damage : float;
var left : boolean = false;
var right : boolean = false;
var strongAgainst : EnemyType = EnemyType.Organic;
var weaponType : WeaponType = WeaponType.Projectile;
var superEffectiveCoef : float = 2;

function Update () {
	CheckStats();
	if (right){
		transform.position += transform.right * speed * Time.deltaTime;
		transform.position.y = 0;

	}
	if (left){
		transform.position -= transform.right * speed * Time.deltaTime;
		transform.position.y = 0;
	}
}

function OnTriggerEnter (other : Collider) {
	if (other.gameObject.GetComponent(Stats) != null){
		if (other.gameObject.GetComponent(Stats).enemyType == strongAgainst){
			other.gameObject.GetComponent(Stats).health -= damage * superEffectiveCoef;
			other.gameObject.GetComponent(Stats).hitWithWeakness = true;
			InstantiateSuperEffectiveSystem(other);
			Debug.Log("SUPER EFFECTIVE");
		} else {
			other.gameObject.GetComponent(Stats).health -= damage;
		}
	} else if (other.transform.parent != null){
		if (other.transform.parent.gameObject.GetComponent(Stats) != null){
			if (other.transform.parent.gameObject.GetComponent(Stats).enemyType == strongAgainst){
				other.transform.parent.gameObject.GetComponent(Stats).health -= damage*superEffectiveCoef;
				other.transform.parent.	gameObject.GetComponent(Stats).hitWithWeakness = true;
				InstantiateSuperEffectiveSystem(other);
				Debug.Log("SUPER EFFECTIVE");
			} else {
				other.transform.parent.gameObject.GetComponent(Stats).health -= damage;
			}
		}
	} else {
		Debug.Log("other and its parents has no stats");
	}
	if (gameObject != null)Destroy (gameObject);
}

function InstantiateSuperEffectiveSystem (other : Collider) {
		var instance : GameObject = Instantiate(other.gameObject.GetComponent(DealDamage).effectiveDebris, 
				transform.position,
				Quaternion.identity);
		instance.transform.parent = other.transform;
		instance.transform.rotation = Quaternion.Euler(transform.rotation.y - 180, 270, 90);
		Debug.Log(instance.name);
		Destroy(instance, 3);
}

function CheckStats () {
	speed = gameObject.GetComponent(Stats).speed;
	damage = gameObject.GetComponent(Stats).damage;
}

File: Aegis/Assets/Scripts/MoveBullet.cs

using UnityEngine;
using System.Collections;

public class MoveBullet : MonoBehaviour, IKillable, ISpeed, ICollisionDamage 
{
    public Vector3 d = Vector3.zero;
    public Vector3 r = Vector3.zero;
    [SerializeField]
    private float damage;
    public float CollisionDamage
    {
        get { return damage; }
        set { damage = value; }
    }
    [SerializeField]
    private float speed;
    public float Speed
    {
        get { return speed; }
        set { speed = value; }
    }
    public bool left = false;
    public bool right = true;

    void Start ()
    {
        rigidbody.AddForce(-transform.right * 100, ForceMode.VelocityChange);
    }

    void FixedUpdate ()
    {
        //rigidbody.AddForce(-transform.right * 100, ForceMode.Force);
    }

    void Update () 
    {
        
	   // if (right)
            
		    //transform.position += transform.right * Speed * Time.deltaTime;
        //Debug.Log(speed);
	  //  if (left)
		   // transform.position -= transform.right * Speed * Time.deltaTime;
    }

    void OnTriggerEnter (Collider other) 
    {
        other.SendMessage("DamageProjectile", CollisionDamage, SendMessageOptions.DontRequireReceiver);
	    if (gameObject != null)Destroy (gameObject);
        //Kill();
    }

    void OnCollisionEnter(Collision other)
    {
        other.gameObject.SendMessage("DamageProjectile", CollisionDamage, SendMessageOptions.DontRequireReceiver);
        //if (gameObject != null) Destroy(gameObject);
        //Kill();
        if (other.gameObject.name == "Shield")
        {
            d = other.transform.position - transform.position;
            r = Vector3.Reflect(d, Vector3.right);
            rigidbody.AddForce(r * 100, ForceMode.VelocityChange);
        }
    }
    
    public void Kill ()
    {
        Destroy(gameObject);
    }
}